core/src/main/java/org/apache/accumulo/core/iterators/ClientIteratorEnvironment.java

```diff
@@ -0,0 +1,220 @@
+      return this;
+    }
+
+    public ClientIteratorEnvironment build() {
+      return new ClientIteratorEnvironment(this);
+    }
+
+  }
+
+  public static final IteratorEnvironment DEFAULT = new Builder().build();
+
+  private final Optional<IteratorScope> scope;
+  private final boolean isFullMajorCompaction;
+  private final Optional<Authorizations> auths;
+  private final boolean isUserCompaction;
+  private final Optional<TableId> tableId;
+  private final Optional<SamplerConfiguration> samplerConfig;
+  private final boolean samplingEnabled;
+  private final Optional<ClientServiceEnvironmentImpl> env;
+
+  private ClientIteratorEnvironment(Builder builder) {
+    this.scope = builder.scope;
+    this.isFullMajorCompaction = builder.isFullMajorCompaction;
+    this.auths = builder.auths;
+    this.isUserCompaction = builder.isUserCompaction;
+    this.tableId = builder.tableId;
+    this.samplerConfig = builder.samplerConfig;
+    this.env = builder.env;
+    this.samplingEnabled = builder.samplingEnabled;
+  }
+
+  /**
+   * Copy constructor used for enabling sample. Only called from {@link cloneWithSamplingEnabled}.
+   */
+  private ClientIteratorEnvironment(ClientIteratorEnvironment copy) {
+    this.scope = copy.scope;
+    this.isFullMajorCompaction = copy.isFullMajorCompaction;
+    this.auths = copy.auths;
+    this.isUserCompaction = copy.isUserCompaction;
+    this.tableId = copy.tableId;
+    this.samplerConfig = copy.samplerConfig;
+    this.env = copy.env;
+    this.samplingEnabled = true;
+  }
+
+  @Override
+  @Deprecated(since = "2.0.0")
+  public SortedKeyValueIterator<Key,Value> reserveMapFileReader(String mapFileName)
+      throws IOException {
+    Configuration hadoopConf = new Configuration();
+    FileSystem fs = FileSystem.get(hadoopConf);
+    return new MapFileIterator(fs, mapFileName, hadoopConf);
+  }
+
+  @Override
+  public IteratorScope getIteratorScope() {
+    return scope.orElseThrow();
+  }
+
+  @Override
+  public boolean isFullMajorCompaction() {
+    if (getIteratorScope() != IteratorScope.majc) {
+      throw new IllegalStateException("Iterator scope is not majc");
+    }
+    return isFullMajorCompaction;
+  }
+
+  @Override
+  @Deprecated(since = "2.0.0")
+  public void registerSideChannel(SortedKeyValueIterator<Key,Value> iter) {
+    throw new UnsupportedOperationException("Feature not supported");
+  }
+
+  @Override
+  public Authorizations getAuthorizations() {
+    if (getIteratorScope() != IteratorScope.scan) {
+      throw new IllegalStateException("Iterator scope is not scan");
+    }
+    return auths.orElseThrow();
+  }
+
+  @Override
+  public IteratorEnvironment cloneWithSamplingEnabled() {
+    String samplerClass =
+        getPluginEnv().getConfiguration(getTableId()).get(Property.TABLE_SAMPLER.getKey());
+    if (samplerClass == null || samplerClass.isBlank()) {
+      throw new SampleNotPresentException();
+    }
```

Comment:
The following seems good for ClientSideIteratorScanner and is tested in `SampleIT.testIterator()`.

```suggestion
    if (samplerConfig.isEmpty()) {
      throw new SampleNotPresentException();
    }
```

Inspecting table props seems like the right way to go, however the bit of code here is not complete for that.  Looking into RFileScanner it may benefit from a secondary check of tables props.  AFIACT there are no tests for calling `cloneWithSamplingEnabled()` when using RFileScanner. In a follow on PR I could add unit test that cover calling this code when using RFileScanner and maybe add some code that inspects the table config as a secondary step when samplerConfig.isEmpty().

Reply:
Opened #5500 to remind myself of the follow on work I want to do here.